### PR TITLE
Discover search URL via service index

### LINKED ISSUE #12038: dotnet tool search command should look up search URL via service index
**Update 6/19/2020:** switch to `SearchQueryService/3.5.0` to avoid unreleased 3.4.0 version.
**Update 6/24/2020:** `SearchQueryService/3.5.0` is available on the [nuget.org service index now](https://api.nuget.org/v3/index.json).

_This only affects .NET CLI from the .NET 5.0 previews since AFAIK these previews are the only place that this code has shipped._

According to the NuGet V3 API documentation for the [Search resource](https://docs.microsoft.com/en-us/nuget/api/search-query-service-resource), clients using the search API should look up the search URL using the search index. Any entry with the `@type` of `SearchQueryService` is a search service. 

The nuget.org V3 service index is: https://api.nuget.org/v3/index.json

In the [`NugetToolSearchApiRequest` class](https://github.com/dotnet/sdk/blob/68baf9d2d992cc53d3c99eed991707e0d071479f/src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs#L48), the search URL is hard coded instead of fetched dynamically from the service index. This is brittle and may break in the future in the URL advertised in the service index changes. Per V3 protocol, package sources are free to change URLs as long as the service index is updated accordingly.

For reference, nuget.org has changed the search URL at least once and the base URL for the sibling resource `RegistrationsBaseUrl` over 5 times.

If this class is able to use the `NuGet.Protocol` NuGet package, I recommend some code like this to discovery the search URL.

```csharp
var repository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
var resource = await repository.GetResourceAsync<ServiceIndexResourceV3>();

var uris = resource.GetServiceEntryUris("SearchQueryService");
foreach (var uri in uris)
{
    Console.WriteLine(uri.AbsoluteUri);
}
```

Output:
```
https://azuresearch-usnc.nuget.org/query
https://azuresearch-ussc.nuget.org/query
```

Two search URLs are advertised by nuget.org. You can choose to use the first only  or take a round robin approach for higher availability like NuGet client code:
https://github.com/NuGet/NuGet.Client/blob/fc9dd52fcdf3c9a299a96de39b641b7c91779b5e/src/NuGet.Core/NuGet.Protocol/Resources/RawSearchResourceV3.cs#L43

Also, we plan on releasing a new resource type (for example `SearchQueryService/3.5.0`) as a signal that a package source supports package type filtering. This is not currently implemented but would ideally go in as part of this change:

```csharp
var uris = resource.GetServiceEntryUris("SearchQueryService/3.5.0");
```

/cc @wli3 @loic-sharma @ryuyu

### PR DESCRIPTION
**Pull request by wli3**
More info in https://github.com/dotnet/sdk/issues/12038
fix https://github.com/dotnet/sdk/issues/12038

### CODE REVIEW

**Comment by wli3** [comment]
@joelverhagen

File: src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs [hunk 1]
```diff
@@ -77,5 +79,14 @@ public async Task<string> GetResult(NugetSearchApiParameter nugetSearchApiParame
 
             return uriBuilder.Uri;
         }
+
+        // More detail on this API https://github.com/dotnet/sdk/issues/12038
+        private static async Task<Uri> DomainAndPath()
+        {
+            var repository = Repository.Factory.GetCoreV3("https://api.nuget.org/v3/index.json");
+            var resource = await repository.GetResourceAsync<ServiceIndexResourceV3>();
+            var uris = resource.GetServiceEntryUris("SearchQueryService");
```
Comment (joelverhagen) [inline src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs:88]: Please use `SearchQueryService/3.5.0`. This marks the service entry that supports package type filtering (which is required by your scenario).

File: src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs [hunk 2]
```diff
@@ -42,10 +44,10 @@ public async Task<string> GetResult(NugetSearchApiParameter nugetSearchApiParame
             return await response.Content.ReadAsStringAsync(cancellation.Token);
         }
 
-        internal static Uri ConstructUrl(string searchTerm = null, int? skip = null, int? take = null,
-            bool prerelease = false)
+        internal static async Task<Uri> ConstructUrl(string searchTerm = null, int? skip = null, int? take = null,
+            bool prerelease = false, Uri domainAndPathOverride = null)
         {
-            var uriBuilder = new UriBuilder("https://azuresearch-usnc.nuget.org/query");
+            var uriBuilder = new UriBuilder(domainAndPathOverride ?? await DomainAndPath());
```
Comment (joelverhagen) [inline src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs:50]: To be clear, this `domainAndPathOverride` is just for unit testing, right? So things can be tested offline?

File: src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs (hunk 1 above)
Comment (joelverhagen) [reply src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs:88]: You could even link to here in your code comment:
https://docs.microsoft.com/en-us/nuget/api/search-query-service-resource#versioning

File: src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs (hunk 2 above)
Comment (wli3) [reply src/Cli/dotnet/NugetSearch/NugetToolSearchApiRequest.cs:50]: yes, for unit testing

**Review by joelverhagen — APPROVED** [review]
Thanks for taking care of this!

**Review by sfoslund — APPROVED** [review]

**Pull request merged**